@vates/disposable/debounceResource.js

```diff
@@ -0,0 +1,55 @@
+const Disposable = require('promise-toolbox/Disposable')
+const { createLogger } = require('@xen-orchestra/log/dist')
+
+const asyncMap = (arrayLike, mapFn, thisArg) => Promise.all(Array.from(arrayLike, mapFn, thisArg))
+
+const { warn } = createLogger('vates:disposable:debounceResource')
+
+exports.createDebounceResource = () => {
+  const flushers = new Set()
+  async function debounceResource(pDisposable, delay = debounceResource.defaultDelay) {
+    if (delay === 0) {
+      return pDisposable
+    }
+
+    const disposable = await pDisposable
+
+    let timeoutId
+    const disposeWrapper = async () => {
+      if (timeoutId !== undefined) {
+        clearTimeout(timeoutId)
+        timeoutId = undefined
+        flushers.delete(flusher)
+
+        try {
+          await disposable.dispose()
+        } catch (error) {
+          warn(error)
+        }
+      }
+    }
+
+    const flusher = () => {
+      const shouldDisposeNow = timeoutId !== undefined
+      if (shouldDisposeNow) {
+        return disposeWrapper()
+      } else {
+        // will dispose ASAP
+        delay = 0
+      }
+    }
+    flushers.add(flusher)
+
+    return new Disposable(disposable.value, () => {
+      timeoutId = setTimeout(disposeWrapper, delay)
+    })
```

Comment:
You can remove the whole `promise-toolbox` dependency by creating the disposable yourself:
```suggestion
    return {
      dispose() {
        timeoutId = setTimeout(disposeWrapper, delay)
      },
      value: disposable.value, 
    }
```

That's the whole point of the disposable API, it's interoperable :slightly_smiling_face:

Reply:
Maybe it should be wrapped with `Promise.resolve`? Because a disposable is a promise.

Reply:
NVM there it's an async function.

Reply:
A disposable is not a promise, it's a “[…] a simple object, containing a value and a dispose function”.